### PR TITLE
drivers: i2c: stm32: Implement PM_DEVICE mode support

### PR DESCRIPTION
**Pull request by fouge**
I2C device can be used as wakeup source when defining the property `wakeup-line` in the I2C node.
Common `STM32_EXTI_LINE_NONE` for declaration and setting of wakeup EXTI line when configured.

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this change.
This looks mostly correct, but scope is too large to fit in one commit.
Please split into several (but self contained) commits

File: drivers/i2c/i2c_ll_stm32.c [hunk 1]
```diff
@@ -223,13 +224,14 @@ static int i2c_stm32_init(const struct device *dev)
...
+	data->clock = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	if (!device_is_ready(data->clock)) {
 		LOG_ERR("clock control device not ready");
 		return -ENODEV;
 	}
 
-	if (clock_control_on(clk,
-		(clock_control_subsys_t *) &cfg->pclken) != 0) {
+	if (clock_control_on(data->clock, (clock_control_subsys_t *)&cfg->pclken) != 0) {
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:234]: Clock changes could be done in a dedicated commit.

File: drivers/serial/uart_stm32.h [hunk 2]
```diff
@@ -16,8 +16,6 @@
 
 #include <stm32_ll_usart.h>
 
-#define STM32_EXTI_LINE_NONE	0xFFFFFFFFU
-
```
Comment (erwango) [inline drivers/serial/uart_stm32.h:20]: This should be moved to a dedicated commit

File: drivers/i2c/i2c_ll_stm32.c [hunk 3]
```diff
@@ -286,6 +303,57 @@ static int i2c_stm32_init(const struct device *dev)
...
+		/* enable clock */
+		err = clock_control_on(data->clock, (clock_control_subsys_t)&config->pclken);
+		if (err != 0) {
+			LOG_ERR("Could not enable I2C clock");
+			return err;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		while (LL_I2C_IsEnabled(config->i2c) == 1) {
+		}
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:330]: This probably works but feels like a hack on `pm_device_action_run`, while PM subsystem expects implementation of `pm_policy_state_lock_put/pm_policy_state_lock_get`.
@ceolin how do you feel with this proposal ?

**Comment by erwango** [comment]
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/37414

File: drivers/i2c/i2c_ll_stm32.c (hunk 3 above)
Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c:330]: I found a better way to handle suspend action depending on driver state and if `CONFIG_PM_DEVICE_RUNTIME` is enabled. I'll push this soon

**Comment by fouge** [comment]
@erwango I rebased the commits

if `PM_DEVICE_ACTION_SUSPEND` is used when I2C device is being used, which happens when I2C waiting for transfer complete for example, the `*_pm_action` returns an error (`-EBUSY`): is it the expected behaviour?

File: drivers/i2c/i2c_ll_stm32.c [hunk 4]
```diff
@@ -283,9 +358,48 @@ static int i2c_stm32_init(const struct device *dev)
...
+		/* Do not suspend if I2C is enabled because it means the driver is waiting for
+		 * an interrupt or a timeout expiration. We can conveniently check the
+		 * I2C state here because the driver disables I2C as soon as it can
+		 * free the peripheral.
+		 * If PM_DEVICE_RUNTIME is enabled, suspend action shouldn't be used until allowed
+		 * in any case.
+		 */
+		if (LL_I2C_IsEnabled(cfg->i2c) != 0) {
+			return -EBUSY;
+		}
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:391]: According to implementation of function `pm_suspend_devices` which will trigger this code, answering `-EBUSY` will end up in a different behavior than if you had used `pm_device_busy_set` or `pm_policy_state_lock_put` in your driver.
See `pm_suspend_devices` implementation: receiving `-EBUSY` will end up returning an error and `pm_resume_devices` will be run, waking up all devices that were previously suspended

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ll_stm32.c (hunk 4 above)
Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c:391]: The PM API is a bit more complex than I thought and I am wondering if I should keep things simple here by only using `pm_device_busy_set` and by removing support for `PM_DEVICE_RUNTIME` at this point. 

What do you think?

Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c]: I finally rebased to use `pm_device_runtime_get`/`pm_device_runtime_put` when `CONFIG_PM_DEVICE_RUNTIME` is enabled and `pm_device_busy_set`/`pm_device_busy_clear` if not

File: drivers/i2c/i2c_ll_stm32.c [hunk 5]
```diff
@@ -347,6 +372,12 @@ static int i2c_stm32_init(const struct device *dev)
 		return ret;
 	}
 
+#if IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:379]: Use `#ifdef CONFIG_PM_DEVICE_RUNTIME`

File: drivers/i2c/i2c_ll_stm32.c [hunk 6]
```diff
@@ -233,6 +233,8 @@ static const struct i2c_driver_api api_funcs = {
 #endif
 };
 
+#ifdef CONFIG_PM_DEVICE
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:236]: Please move this directly to the commit where you introduce `i2c_stm32_suspend`

File: drivers/i2c/i2c_ll_stm32.c [hunk 7]
```diff
@@ -53,9 +56,23 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 	data->dev_config = config;
 
 	k_sem_take(&data->bus_mutex, K_FOREVER);
+
+#if IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)
+	(void)pm_device_runtime_get(dev);
+#else
+	pm_device_busy_set(dev);
+#endif
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:64]: Should be:
```
#ifdef CONFIG_PM_DEVICE_RUNTIME
	(void)pm_device_runtime_get(dev);
#else
	pm_device_busy_set(dev);
#endif
```

Though, I wonder if the #idef is actually needed.

Another option would be:
```
	(void)pm_device_runtime_get(dev);
	pm_device_busy_set(dev);
```
This should compile in all combinations and make the code smoother to read.
There might be an overhead but I don't think it is killing.

@ceolin, @gmarull do you have any advice on the preferred implementation if we aim to implement `PM_DEVICE_RUNTIME` support on such driver ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ll_stm32.c (hunk 7 above)
Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c:64]: I thought about it but then wondered if we should call `pm_device_busy_set` with `CONFIG_PM_DEVICE_RUNTIME` so that's why I have either one or this other.

File: drivers/i2c/i2c_ll_stm32.c (hunk 5 above)
Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c:379]: fixed (rebased)

File: drivers/i2c/i2c_ll_stm32.c (hunk 7 above)
Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c:64]: @ceolin @gmarull any advice?

File: drivers/i2c/i2c_ll_stm32.c (hunk 6 above)
Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c:236]: done

File: drivers/i2c/i2c_ll_stm32.c (hunk 7 above)
Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c]: @fouge Have you been able to verify that both configurations work as expected ?

Comment (fouge) [reply drivers/i2c/i2c_ll_stm32.c]: It runs but I am not able to measure any significant drop in current consumption.

I obtain a baseline of ~90uA overall when using `PM_DEVICE` on my system, and 1.25mA with `PM_DEVICE_RUNTIME`, which is likely due to some other factors.

Comment (erwango) [reply drivers/i2c/i2c_ll_stm32.c]: Did you try to prevent core going to stop/sleep states ?

File: drivers/i2c/i2c_ll_stm32.c [hunk 8]
```diff
@@ -53,9 +56,23 @@ int i2c_stm32_runtime_configure(const struct device *dev, uint32_t config)
 	data->dev_config = config;
 
 	k_sem_take(&data->bus_mutex, K_FOREVER);
+
+#ifdef CONFIG_PM_DEVICE_RUNTIME
+	(void)pm_device_runtime_get(dev);
+#else
+	pm_device_busy_set(dev);
+#endif
```
Comment (erwango) [inline drivers/i2c/i2c_ll_stm32.c:64]: @ceolin Is the the expected code when both PM configurations have to be supported ?

**Comment by erwango** [comment]
> I am wondering if we should remove support for the wake-up line. It hasn't been correctly tested and I think can only be added by a few microcontroller series such as the STM32L1x.

Sure, this can be added later

**Comment by fouge** [comment]
Also, since I need to rebase and looking at #53216 that has been merged, it seems like we would need to add the "clocks-name" property to the I2C node so that we can differentiate clocks used by default and clocks used when in sleep mode. We would end up with "default" clocks and "sleep" clocks (mostly L1 series I guess).
This way, we can turn on and off clocks depending on their names.
What do you think @erwango? 
cc @gautierg-st

**Comment by erwango** [comment]
> Also, since I need to rebase and looking at #53216 that has been merged, it seems like we would need to add the "clocks-name" property to the I2C node so that we can differentiate clocks used by default and clocks used when in sleep mode. We would end up with "default" clocks and "sleep" clocks (mostly L1 series I guess). This way, we can turn on and off clocks depending on their names. What do you think @erwango? cc @gautierg-st

Let's get this first part merged and we'll see that afterwards.

**Review by erwango — DISMISSED** [review]

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]
